lib/stream_data.ex

```diff
@@ -701,20 +722,48 @@ defmodule StreamData do
   #       |> LazyTree.flatten()
   #     end)
   #
-  @spec list_of(t(a)) :: t([a]) when a: term
-  def list_of(data) do
+  @spec list_of(t(a), keyword) :: t([a]) when a: term
+  def list_of(data, options \\ []) do
+    list_length_range_fun = list_length_range_fun(Keyword.take(options, [:length, :min_length, :max_length]))
+
     new(fn seed, size ->
       {seed1, seed2} = split_seed(seed)
-      length = uniform_in_range(0..size, seed1)
+      min_length.._ = length_range = list_length_range_fun.(size)
+      length = uniform_in_range(length_range, seed1)
 
       data
       |> call_n_times(seed2, size, length, [])
       |> LazyTree.zip()
-      |> LazyTree.map(&list_lazy_tree/1)
+      |> LazyTree.map(&list_lazy_tree(&1, min_length))
       |> LazyTree.flatten()
     end)
   end
 
+  defp list_length_range_fun(options) do
+    {min, max} =
+      case options[:length] do
+        length when is_integer(length) and length >= 0 ->
+          {length, length}
+        min..max when min >= 0 and max >= 0 ->
+          {min(min, max), max(min, max)}
+        nil ->
+          min_length = options[:min_length] || 0
+          max_length = options[:max_length] || :infinity
+          unless is_integer(min_length) and min_length >= 0 do
+            raise ArgumentError, ":min_length must be a positive integer, got: #{inspect(min_length)}"
+          end
+          unless (is_integer(max_length) and max_length >= 0) or max_length == :infinity do
+            raise ArgumentError, ":max_length must be a positive integer, got: #{inspect(max_length)}"
+          end
+          {min_length, max_length}
+        other ->
+          raise ArgumentError, ":length must be a positive integer or a range " <>
+                               "of positive integers, got: #{inspect(other)}"
+      end
+
+    &(min..(max |> min(_size = &1) |> max(min)))
```

Comment:
I find those pipes very unclear (because the function names and variable names). Which case is it trying to handle?